zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala

```diff
@@ -389,27 +313,33 @@ class CassieSpanStore(
   ): Future[Seq[IndexedTraceId]] = {
     QueryGetTraceIdsByAnnotationCounter.incr()
     val key = annotationKey(serviceName, annotation, value)
-    AnnotationsIndex.getRowSlice(key, None, Some(endTs), limit, Order.Reversed) map colToIndexedTraceId
+
+    pool {
+      repository
+        .getTraceIdsByAnnotation(annotationKey(serviceName, annotation, value), endTs, limit)
+        .map { case (traceId :java.lang.Long, ts :java.lang.Long) => IndexedTraceId(traceId, timestamp = ts) }
+        .toSeq
+    }
   }
 
   def getTracesDuration(traceIds: Seq[Long]): Future[Seq[TraceIdDuration]] = {
     QueryGetTracesDurationStat.add(traceIds.size)
 
-    val traceIdSet = traceIds.toSet.asJava
-
-    Future.collect(Seq(
-      DurationIndex.multigetRows(traceIdSet, None, None, Order.Normal, 1),
-      DurationIndex.multigetRows(traceIdSet, None, None, Order.Reversed, 1)
-    )) map { results =>
-      val Seq(startRows, endRows) = results map { rows =>
-        rows.asScala.toSeq map { case (traceId, cols) =>
-          cols.asScala.headOption map { case (_, col) => (traceId, col.name) }
-        }
-      }
-      startRows zip(endRows) collect {
-        case (Some((startId, startTs)), Some((endId, endTs))) if (startId == endId) =>
-          TraceIdDuration(endId, endTs - startTs, startTs)
+    val traceIdSet = traceIds.toArray
+
+    pool {
+      (repository.getTraceDuration(true, traceIdSet)
+      .map { case (traceId :java.lang.Long, ts :java.lang.Long) => (traceId, ("start", ts))}
+      .toSeq ++
+        (repository.getTraceDuration(false, traceIdSet)
+        .map { case (traceId :java.lang.Long, ts :java.lang.Long) => (traceId, ("end", ts))}
+        .toSeq))
+      .groupBy(_._1)
+      .mapValues(_.map(_._2))
```

Comment:
i played around with `.mapValues(_.values)` and `.flatten` and found nothing that worked :-(

i'm open to working suggestions otherwise.